src/uart.c

```diff
@@ -40,3 +41,106 @@ struct metal_uart *metal_uart_get_device(unsigned int device_num) {
 
     return NULL;
 }
+
+void _metal_uart_async_callback(int id, void *priv) {
+    struct metal_uart *uart = (struct metal_uart *)priv;
+
+    if (uart->_tx_async.in_progress && metal_uart_txready(uart)) {
+        while (metal_uart_txready(uart)) {
+            metal_uart_putc(uart, uart->_tx_async.buf[uart->_tx_async.index]);
+
+            uart->_tx_async.index += 1;
+            if (uart->_tx_async.index >= uart->_tx_async.len) {
+                uart->_tx_async.in_progress = 0;
+                uart->_tx_async.cb(uart, METAL_UART_TX_DONE);
+                metal_uart_transmit_interrupt_disable(uart);
+                break;
+            }
+        }
+    }
+
+    if (uart->_rx_async.in_progress) {
+        while (1) {
+            int c;
+
+            metal_uart_getc(uart, &c);
+            if (c == -1) {
+                /* Nothing to read */
+                break;
+            }
+
+            uart->_rx_async.buf[uart->_rx_async.index] = (char)c;
+
+            uart->_rx_async.index += 1;
+            if (uart->_rx_async.index >= uart->_rx_async.len) {
+                uart->_rx_async.in_progress = 0;
+                uart->_rx_async.cb(uart, METAL_UART_RX_DONE);
+                metal_uart_receive_interrupt_disable(uart);
+                break;
+            }
+        }
+    }
+}
+
+int metal_uart_send_async(struct metal_uart *uart, char *buf, size_t len,
+                          metal_uart_callback_t callback) {
+    if (metal_uart_send_async_busy(uart)) {
+        return 1;
+    }
+
+    uart->_tx_async.in_progress = 1;
+    uart->_tx_async.buf = buf;
+    uart->_tx_async.len = len;
+    uart->_tx_async.index = 0;
+    uart->_tx_async.cb = callback;
+
+    /* Register _metal_uart_async_callback as the interrupt handler for the
+     * uart interrupt */
+    struct metal_interrupt *uart_intr = metal_uart_interrupt_controller(uart);
+    int uart_id = metal_uart_get_interrupt_id(uart);
```

Comment:
Could be done in initialization function ... with pointers on parameters to give back uart_intr, uart_d to caller ... or having internal (to driver) parameters to save them ?

Reply:
You're thinking of some kind of `metal_uart_async_init()`?

Reply:
No, I was thinking of more general initialization function ... that covers the whole driver/IP aspect ... in order to provide "one entry point" to user ...
Of course, it may be convenient to have feature specific functions. But both approaches can "live together", can't they ?